ddl/schema.go

```diff
@@ -257,6 +257,124 @@ func onDropSchema(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	ver, err = updateSchemaVersion(t, job)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	switch schemaInfo.State {
+	case model.StateNone:
+		// none -> write only
+		// check GC enable and update flag.
+		if gcEnable {
+			job.Args[checkFlagIndexInJobArgs] = recoverCheckFlagEnableGC
+		} else {
+			job.Args[checkFlagIndexInJobArgs] = recoverCheckFlagDisableGC
+		}
+		err := checkSchemaNotExists(d, t, schemaInfo.ID, schemaInfo)
+		if err != nil {
+			if infoschema.ErrDatabaseExists.Equal(err) {
+				// The database already exists, can't create it, we should cancel this job now.
+				job.State = model.JobStateCancelled
+			}
+			return ver, errors.Trace(err)
+		}
+		err = t.CreateDatabase(schemaInfo)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		schemaInfo.State = model.StateWriteOnly
+		err = t.UpdateDatabase(schemaInfo)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StateWriteOnly
+	case model.StateWriteOnly:
+		// write only -> public
+		// do recover table.
+		if gcEnable {
+			err = disableGC(w)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, errors.Errorf("disable gc failed, try again later. err: %v", err)
+			}
+		}
+		// check GC safe point
+		err = checkSafePoint(w, snapshotTS)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+		job.SchemaID = schemaInfo.ID
+		for _, recoverInfo := range recoversInfo {
+			var tids []int64
+			if recoverInfo.TableInfo.GetPartitionInfo() != nil {
+				tids = getPartitionIDs(recoverInfo.TableInfo)
+				tids = append(tids, recoverInfo.TableInfo.ID)
+			} else {
+				tids = []int64{recoverInfo.TableInfo.ID}
+			}
+			tableRuleID, partRuleIDs, oldRuleIDs, oldRules, err := getOldLabelRules(recoverInfo.TableInfo, oldSchemaName, recoverInfo.TableInfo.Name.L)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, errors.Wrapf(err, "failed to get old label rules from PD")
+			}
+			// Remove dropped table DDL job from gc_delete_range table.
+			err = w.delRangeManager.removeFromGCDeleteRange(w.ddlJobCtx, dropJobID, tids)
+			if err != nil {
+				return ver, errors.Trace(err)
+			}
+			recoverInfo.TableInfo.State = model.StatePublic
+			recoverInfo.TableInfo.UpdateTS = t.StartTS
+			err = t.CreateTableAndSetAutoID(schemaInfo.ID, recoverInfo.TableInfo, recoverInfo.AutoIDs.RowID, recoverInfo.AutoIDs.RandomID)
+			if err != nil {
+				return ver, errors.Trace(err)
+			}
+			err = updateLabelRules(job, recoverInfo.TableInfo, oldRules, tableRuleID, partRuleIDs, oldRuleIDs, recoverInfo.TableInfo.ID)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, errors.Wrapf(err, "failed to update the label rule to PD")
+			}
+			ver, err = updateVersionAndTableInfo(t, job, recoverInfo.TableInfo, false)
+			if err != nil {
+				return ver, errors.Trace(err)
+			}
```

Comment:
```suggestion
```

Reply:
Should only update schema version once in this situation. the related schema change should be set in `updateSchemaVersion`, related with the comment:  https://github.com/pingcap/tidb/pull/30217/files#r767778383